src/server/DeviceServer.js

```diff
@@ -584,15 +616,167 @@ class DeviceServer {
         messageName,
         device.onDeviceEvent,
         {
-          connectionID: isSystemEvent ? device.getConnectionKey() : null,
-          mydevices: isFromMyDevices,
-          userID: ownerID,
+          filterOptions: {
+            connectionID: isSystemEvent ? device.getConnectionKey() : null,
+            mydevices: isFromMyDevices,
+            userID: ownerID,
+          },
+          subscriberID: deviceID,
         },
-        deviceID,
       );
     });
   };
 
+  _onSparkServerCallDeviceFunctionRequest = async (event: Event): Promise<void> => {
+    const {
+      deviceID,
+      functionArguments,
+      functionName,
+      responseEventName,
+    } = nullthrows(event.context);
+    try {
+      const device = this.getDevice(deviceID);
+      if (!device) {
+        throw new Error('Could not get device for ID');
+      }
+
+      this._eventPublisher.publish({
```

Comment:
Ideally these would be `nextTick` so it does't do all the synchronous callbacks.